File: types/backtest.ts
/**
 * Vysted Terminal — backtest engine wire contract.
 *
 * Phase 4 ships a custom event-driven backtest engine in the Python sidecar
 * (`services/backtest_engine.py`). Strategies implement a `BacktestStrategy`
 * ABC; the engine fetches historical OHLCV through Phase 1's
 * `provider_registry` (yfinance + ccxt + openbb-mcp), walks bars in order,
 * captures trades into a trade log, computes equity curve + drawdown +
 * Sharpe + win rate at end. Walk-forward is supported via slice-and-aggregate.
 *
 * The AI Strategy Critic agent (`sidecar/agents/strategy_critic.json`) has
 * `backtest_summary` in its tool list since v0.4.0 — Phase 4 wires that tool
 * to a digest of `BacktestResult` so the critic can interrogate the result
 * end-to-end (Use Case 2 in BLUEPRINT §10).
 *
 * Why a custom engine instead of vectorbt/backtrader (both common):
 * - backtrader development stopped in 2018; security maintenance is unclear.
 * - vectorbt pulls heavy deps (numba) that risk pushing the main sidecar
 *   `--onefile` binary past the 120 MB threshold (CLAUDE.md Phase-3
 *   Gotcha — current main 67 MB).
 * - BLUEPRINT §7 Phase 4 wording "vectorbt+backtrader patterns" supports
 *   drawing on their design ideas without runtime dependency.
 */

// ---------------------------------------------------------------------------
// Strategy spec
// ---------------------------------------------------------------------------

/**
 * A backtestable strategy registered in the sidecar's strategy registry.
 * Strategies live server-side as Python `BacktestStrategy` subclasses
 * (`services/backtest_strategies.py`); the wire surface is this metadata
 * plus the JSON-Schema-validated `params` shape.
 */
export interface BacktestStrategySpec {
  /** Stable identifier, e.g. `"mean_reversion"`, `"trend_following"`. */
  id: string;
  /** Display name shown in the strategy picker. */
  name: string;
  /** One-line description shown beneath the name. */
  description: string;
  /**
   * JSON Schema (draft-07 subset) describing the `params` shape this
   * strategy accepts. The frontend renders a form from this schema.
   */
  paramsSchema: Record<string, unknown>;
}

// ---------------------------------------------------------------------------
// Run request
// ---------------------------------------------------------------------------

/** Fee + slippage model applied to every simulated fill. */
export interface BacktestFeeModel {
  /** Per-side commission in basis points (5 = 0.05%). Default `5` for equities. */
  feeBps: number;
  /** Per-side slippage in basis points (5 = 0.05%). Default `5` for equities. */
  slippageBps: number;
}

/** `POST /backtest/run` request body. */
export interface BacktestRequest {
  /** Registered strategy id. */
  strategyId: string;
  /** Strategy-specific params; validated against `BacktestStrategySpec.paramsSchema`. */
  params: Record<string, unknown>;
  /** Tickers to trade. Many strategies trade only `symbols[0]`. */
  symbols: string[];
  /** Inclusive ISO-8601 date `YYYY-MM-DD`. */
  startDate: string;
  /** Inclusive ISO-8601 date `YYYY-MM-DD`. */
  endDate: string;
  /** Starting cash, in account currency. Default 100_000. */
  initialCapital: number;
  /** Fee model; default is equity defaults if omitted. */
  feeModel?: BacktestFeeModel;
  /** Walk-forward slice count (`1` = single in-sample run, default). */
  walkForwardSlices?: number;
}

// ---------------------------------------------------------------------------
// Result
// ---------------------------------------------------------------------------

/** A single trade in the trade log. */
export interface BacktestTrade {
  /** Stable id per trade. */
  id: string;
  symbol: string;
  side: "buy" | "sell";
  /** ISO timestamp of fill. */
  enteredAt: string;
  exitedAt?: string;
  entryPrice: number;
  exitPrice?: number;
  quantity: number;
  /** Realized P&L in account currency; undefined while the trade is open. */
  pnl?: number;
  /** Reason the trade closed (`"strategy"`, `"stop-loss"`, `"end-of-data"`). */
  closeReason?: string;
}

/** One point on the equity curve, sampled per bar. */
export interface EquityCurvePoint {
  /** ISO timestamp of the bar. */
  timestamp: string;
  /** Total account equity including open-position mark-to-market. */
  equity: number;
  /** Drawdown from running peak, expressed as a negative percent (-0.12 = -12%). */
  drawdownPct: number;
}

/** Walk-forward slice summary — one entry per slice when slices > 1. */
export interface WalkForwardSlice {
  index: number;
  startDate: string;
  endDate: string;
  totalReturn: number;
  sharpe: number;
  trades: number;
}

/** Aggregated metrics for a complete backtest. */
export interface BacktestMetrics {
  totalReturn: number;
  annualizedReturn: number;
  /** Annualised Sharpe (252-day, risk-free assumed 0 for simplicity in v0.5.0). */
  sharpe: number;
  sortino: number;
  calmar: number;
  /** Peak-to-trough drawdown as a negative fraction (-0.25 = -25%). */
  maxDrawdownPct: number;
  /** Win rate as fraction (0.55 = 55%). */
  winRate: number;
  /** Number of closed trades. */
  tradeCount: number;
  /** Best single-trade P&L. */
  bestTradePnl: number;
  /** Worst single-trade P&L. */
  worstTradePnl: number;
}

/** Full backtest result returned by `GET /backtest/runs/{id}`. */
export interface BacktestResult {
  runId: string;
  strategyId: string;
  request: BacktestRequest;
  metrics: BacktestMetrics;
  trades: BacktestTrade[];
  equityCurve: EquityCurvePoint[];
  walkForwardSlices?: WalkForwardSlice[];
  startedAt: number;
  durationMs: number;
}

// ---------------------------------------------------------------------------
// Run event stream
// ---------------------------------------------------------------------------

/** One event in a backtest run's SSE stream. */
export type BacktestRunEvent =
  | { kind: "run-start"; runId: string; totalBars: number; startedAt: number }
  | { kind: "progress"; runId: string; barsProcessed: number; equity: number }
  | { kind: "trade"; runId: string; trade: BacktestTrade }
  | { kind: "run-complete"; runId: string; result: BacktestResult }
  | { kind: "run-error"; runId: string; message: string };

// ---------------------------------------------------------------------------
// Strategy Critic backtest_summary tool digest
// ---------------------------------------------------------------------------

/**
 * The compact digest the Strategy Critic agent's `backtest_summary` tool
 * returns. The agent does NOT receive the raw equity curve (thousands of
 * points) or full trade log — only what's needed to apply the 9-section
 * critique framework. This shape stays stable across releases so the agent
 * prompt does not drift.
 */
export interface BacktestSummary {
  runId: string;
  strategyId: string;
  strategyParams: Record<string, unknown>;
  symbols: string[];
  startDate: string;
  endDate: string;
  metrics: BacktestMetrics;
  /** Up to 20 most-recent trades for sample-size reasoning. */
  recentTrades: BacktestTrade[];
  /** Top-3 and bottom-3 trades by P&L for outlier reasoning. */
  bestTrades: BacktestTrade[];
  worstTrades: BacktestTrade[];
  walkForwardSlices?: WalkForwardSlice[];
}

File: sidecar/models/backtest.py
"""Backtest engine Pydantic models — mirror of ``types/backtest.ts``.

Phase 4 ships a custom event-driven backtest engine
(``services/backtest_engine.py``). Strategies subclass ``BacktestStrategy``
ABC; the engine fetches historical OHLCV via Phase 1's ``provider_registry``,
walks bars, captures trades, computes metrics + equity curve + drawdown.

The ``BacktestSummary`` shape is the Strategy Critic agent's
``backtest_summary`` tool's return shape — kept compact so the agent prompt
does not drift across releases.
"""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field


class BacktestStrategySpec(BaseModel):
    """A registered backtest strategy's metadata."""

    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    id: str
    name: str
    description: str
    params_schema: dict[str, Any] = Field(alias="paramsSchema")


class BacktestFeeModel(BaseModel):
    """Fee + slippage model applied to every simulated fill."""

    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    fee_bps: float = Field(alias="feeBps", default=5.0)
    slippage_bps: float = Field(alias="slippageBps", default=5.0)


class BacktestRequest(BaseModel):
    """``POST /backtest/run`` request body."""

    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    strategy_id: str = Field(alias="strategyId")
    params: dict[str, Any] = Field(default_factory=dict)
    symbols: list[str]
    start_date: str = Field(alias="startDate")
    end_date: str = Field(alias="endDate")
    initial_capital: float = Field(alias="initialCapital", default=100_000.0)
    fee_model: BacktestFeeModel | None = Field(default=None, alias="feeModel")
    walk_forward_slices: int = Field(alias="walkForwardSlices", default=1, ge=1, le=10)


class BacktestTrade(BaseModel):
    """A single trade in the trade log."""

    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    id: str
    symbol: str
    side: Literal["buy", "sell"]
    entered_at: str = Field(alias="enteredAt")
    exited_at: str | None = Field(default=None, alias="exitedAt")
    entry_price: float = Field(alias="entryPrice")
    exit_price: float | None = Field(default=None, alias="exitPrice")
    quantity: float
    pnl: float | None = None
    close_reason: str | None = Field(default=None, alias="closeReason")


class EquityCurvePoint(BaseModel):
    """One point on the equity curve, sampled per bar."""

    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    timestamp: str
    equity: float
    drawdown_pct: float = Field(alias="drawdownPct")


class WalkForwardSlice(BaseModel):
    """Walk-forward slice summary."""

    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    index: int
    start_date: str = Field(alias="startDate")
    end_date: str = Field(alias="endDate")
    total_return: float = Field(alias="totalReturn")
    sharpe: float
    trades: int


class BacktestMetrics(BaseModel):
    """Aggregated metrics for a complete backtest."""

    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    total_return: float = Field(alias="totalReturn")
    annualized_return: float = Field(alias="annualizedReturn")
    sharpe: float
    sortino: float
    calmar: float
    max_drawdown_pct: float = Field(alias="maxDrawdownPct")
    win_rate: float = Field(alias="winRate")
    trade_count: int = Field(alias="tradeCount")
    best_trade_pnl: float = Field(alias="bestTradePnl")
    worst_trade_pnl: float = Field(alias="worstTradePnl")


class BacktestResult(BaseModel):
    """Full backtest result returned by ``GET /backtest/runs/{id}``."""

    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    run_id: str = Field(alias="runId")
    strategy_id: str = Field(alias="strategyId")
    request: BacktestRequest
    metrics: BacktestMetrics
    trades: list[BacktestTrade]
    equity_curve: list[EquityCurvePoint] = Field(alias="equityCurve")
    walk_forward_slices: list[WalkForwardSlice] | None = Field(
        default=None, alias="walkForwardSlices"
    )
    started_at: int = Field(alias="startedAt")
    duration_ms: float = Field(alias="durationMs")


class BacktestRunEvent(BaseModel):
    """One SSE event in a backtest run."""

    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    kind: Literal["run-start", "progress", "trade", "run-complete", "run-error"]
    run_id: str = Field(alias="runId")
    total_bars: int | None = Field(default=None, alias="totalBars")
    started_at: int | None = Field(default=None, alias="startedAt")
    bars_processed: int | None = Field(default=None, alias="barsProcessed")
    equity: float | None = None
    trade: BacktestTrade | None = None
    result: BacktestResult | None = None
    message: str | None = None


class BacktestSummary(BaseModel):
    """Strategy Critic ``backtest_summary`` tool's return shape.

    Compact digest — the agent does NOT receive the raw equity curve or
    full trade log, only what's needed to apply the 9-section critique
    framework from ``sidecar/agents/strategy_critic.json``.
    """

    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    run_id: str = Field(alias="runId")
    strategy_id: str = Field(alias="strategyId")
    strategy_params: dict[str, Any] = Field(alias="strategyParams")
    symbols: list[str]
    start_date: str = Field(alias="startDate")
    end_date: str = Field(alias="endDate")
    metrics: BacktestMetrics
    recent_trades: list[BacktestTrade] = Field(alias="recentTrades")
    best_trades: list[BacktestTrade] = Field(alias="bestTrades")
    worst_trades: list[BacktestTrade] = Field(alias="worstTrades")
    walk_forward_slices: list[WalkForwardSlice] | None = Field(
        default=None, alias="walkForwardSlices"
    )
